util/types/json/path_expr.go

```diff
@@ -0,0 +1,133 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package json
+
+import (
+	"regexp"
+	"strconv"
+
+	"github.com/juju/errors"
+)
+
+/*
+	From MySQL 5.7, JSON path expression grammar:
+		pathExpression ::= scope pathLeg (pathLeg)*
+		scope ::= [ columnReference ] '$'
+		columnReference ::= // omit...
+		pathLeg ::= member | arrayLocation | '**'
+		member ::= '.' (keyName | '*')
+		arrayLocation ::= '[' (non-negative-integer | '*') ']'
+		keyName ::= ECMAScript-identifier | ECMAScript-string-literal
+
+	And some implemetion limits in MySQL 5.7:
+		1) columnReference in scope must be empty now;
+		2) double asterisk(**) could not be last leg;
+
+	Examples:
+		select json_extract('{"a": "b", "c": [1, "2"]}', '$.a') -> "b"
+		select json_extract('{"a": "b", "c": [1, "2"]}', '$.c') -> [1, "2"]
+		select json_extract('{"a": "b", "c": [1, "2"]}', '$.a', '$.c') -> ["b", [1, "2"]]
+		select json_extract('{"a": "b", "c": [1, "2"]}', '$.c[0]') -> 1
+		select json_extract('{"a": "b", "c": [1, "2"]}', '$.c[2]') -> NULL
+		select json_extract('{"a": "b", "c": [1, "2"]}', '$.c[*]') -> [1, "2"]
+		select json_extract('{"a": "b", "c": [1, "2"]}', '$.*') -> ["b", [1, "2"]]
+	TODO:
+		1) add double asterisk support;
+*/
+var blankRe = regexp.MustCompile(`\s`)
+var jsonPathExprLegRe = regexp.MustCompile(`(\.([a-zA-Z_][a-zA-Z0-9_]*|\*)|(\[([0-9]+|\*)\]))`)
+
+// pathLeg is only used by PathExpression.
+type pathLeg struct {
+	start        int  // start offset of the leg in raw string, inclusive.
+	end          int  // end offset of the leg in raw string, exclusive.
+	isArrayIndex bool // the leg is an array index or not.
+	arrayIndex   int  // if isArrayIndex is true, the value shoud be parsed into here.
```

Comment:
I'm not sure will this be better, we can define a constant like `notArrayIndex` which is set to 2.
thus we can save the storage spaces of isArrayIndex.
`arrayIndex == -2`  ==> `isArrayIndex is false`
`arrayIndex == -1` ==> `isArrayIndex == true && arrayIndexAsterisk`
`arrayIndex >= 0`  ==> `isArrayIndex == true`